File: 04-structural-patterns/05-facade/05-facade.go
package main

import (
	"database/sql"
	"fmt"
	"log"

	_ "modernc.org/sqlite"
)

// Subsystem: DB Connection
type DBConnection struct{}

func (db *DBConnection) Connect() *sql.DB {
	conn, err := sql.Open("sqlite", ":memory:")
	if err != nil {
		log.Fatal(err)
	}
	return conn
}

// Subsystem: Product Operations
type ProductRepository struct {
	conn *sql.DB
}

func NewProductRepository(conn *sql.DB) *ProductRepository {
	_, _ = conn.Exec(`CREATE TABLE products (id INTEGER, name TEXT, price REAL)`)
	return &ProductRepository{conn: conn}
}

func (r *ProductRepository) AddProduct(id int, name string, price float64) {
	_, _ = r.conn.Exec(`INSERT INTO products VALUES (?, ?, ?)`, id, name, price)
}

func (r *ProductRepository) ListProducts() {
	rows, _ := r.conn.Query(`SELECT id, name, price FROM products`)
	defer rows.Close()
	for rows.Next() {
		var id int
		var name string
		var price float64
		_ = rows.Scan(&id, &name, &price)
		fmt.Printf("ID: %d, Name: %s, Price: %.2f\n", id, name, price)
	}
}

// Facade
type ProductService struct {
	db   *sql.DB
	repo *ProductRepository
}

func NewProductService() *ProductService {
	conn := (&DBConnection{}).Connect()
	repo := NewProductRepository(conn)
	return &ProductService{db: conn, repo: repo}
}

func (s *ProductService) AddProduct(id int, name string, price float64) {
	fmt.Println("Validating product...")
	s.repo.AddProduct(id, name, price)
	fmt.Println("Product added.")
}

func (s *ProductService) ListAll() {
	s.repo.ListProducts()
}

// Client
func main() {
	service := NewProductService()
	service.AddProduct(1, "Laptop", 1200)
	service.AddProduct(2, "Mouse", 25)

	service.ListAll()
}

File: 04-structural-patterns/05-facade/05-facade.py

import sqlite3

# Subsystem: DB Connection
class DBConnection:
    def connect(self):
        return sqlite3.connect(":memory:")

# Subsystem: Product Operations
class ProductRepository:
    def __init__(self, conn):
        self.conn = conn
        self.create_table()

    def create_table(self):
        self.conn.execute("CREATE TABLE products (id INTEGER, name TEXT, price REAL)")

    def add_product(self, id, name, price):
        self.conn.execute("INSERT INTO products VALUES (?, ?, ?)", (id, name, price))
        self.conn.commit()

    def list_products(self):
        return self.conn.execute("SELECT * FROM products").fetchall()

# Facade
class ProductService:
    def __init__(self):
        self.db = DBConnection()
        self.conn = self.db.connect()
        self.repo = ProductRepository(self.conn)

    def add_product(self, id, name, price):
        print("Validating product...")
        self.repo.add_product(id, name, price)
        print("Product added.")

    def list_all(self):
        return self.repo.list_products()

# Client
service = ProductService()
service.add_product(1, "Laptop", 1200)
service.add_product(2, "Mouse", 25)

print(service.list_all())